torch/fft/__init__.py

```diff
@@ -549,3 +555,186 @@
+        The spacing between individual samples of the FFT input.
+        The default assumes unit spacing, dividing that result by the actual
+        spacing gives the result in physical frequency units.
+
+Keyword Args:
+    {dtype}
+    {layout}
+    {device}
+    {requires_grad}
+
+Example:
+
+    >>> import torch.fft
+    >>> torch.fft.fftfreq(5)
+    tensor([ 0.0000,  0.2000,  0.4000, -0.4000, -0.2000])
+
+    For even input, we can see the Nyquist frequency at ``f[2]`` is given as
+    negative:
+
+    >>> torch.fft.fftfreq(4)
+    tensor([ 0.0000,  0.2500, -0.5000, -0.2500])
+
+""".format(**factory_common_args))
+
+rfftfreq = _add_docstr(_fft.fft_rfftfreq, r"""
+rfftfreq(n, d=1.0, *, dtype=None, layout=torch.strided, device=None, requires_grad=False) -> Tensor
+
+Computes the sample frequencies for :func:`~torch.fft.rfft` with a signal of size :attr:`n`.
+
+Note:
+
+    :func:`~torch.fft.rfft` returns Hermitian one-sided output, so only the
+    positive frequency terms are returned. For a real FFT of length :attr:`n`
+    and with inputs spaced in length unit :attr:`d`, the frequencies are::
+
+        f = torch.arange((n + 1) // 2) / (d * n)
+
+Note:
+    For even lengths, the Nyquist frequency at ``f[n/2]`` can be thought of as
+    either negative or positive. Unlike :func:`~torch.fft.fftfreq`,
+    :func:`~torch.fft.rfftfreq` always returns is as positive.
+
+
+Args:
+    n (int): the real FFT length
+    d (float, optional): The sampling length scale.
+        The spacing between individual samples of the FFT input.
+        The default assumes unit spacing, dividing that result by the actual
+        spacing gives the result in physical frequency units.
+
+Keyword Args:
+    {dtype}
+    {layout}
+    {device}
+    {requires_grad}
+
+Example:
+
+    >>> import torch.fft
+    >>> torch.fft.rfftfreq(5)
+    tensor([ 0.0000,  0.2000,  0.4000])
+
+    >>> torch.fft.rfftfreq(4)
+    tensor([ 0.0000,  0.2500, 0.5000])
+
+    Compared to the output from :func:`~torch.fft.fftfreq`, we see that the
+    Nyquist frequency at ``f[2]`` has changed sign:
+    >>> torch.fft.fftfreq(4)
+    tensor([ 0.0000,  0.2500, -0.5000, -0.2500])
+
+""".format(**factory_common_args))
+
+fftshift = _add_docstr(_fft.fft_fftshift, r"""
+fftshift(input, dim=None) -> Tensor
+
+Reorder n-dimensional FFT output to have negative frequency terms first. 
+
+Note:
+    By convention, the FFT returns positive frequency terms first, followed by
+    the negative frequencies in reverse order, so that ``f[-i]`` for all
+    :math:`i \leq n/2` in Python gives the negative frequency terms.
+    :func:`~torch.fft.fftshift` rearranges all frequencies into ascending order
+    from negative to positive with the zero-frequency term in the center.
+
+Note:
+    For even lengths, the Nyquist frequency at ``f[n/2]`` can be thought of as
+    either negative or positive. Like :func:`~torch.fft.fftfreq`,
+    :func:`~torch.fft.fftshift` always considers it to be negative.
```

Comment:
This note is a little confusing because fftshift doesn't actually look at its input. Isn't it more like the function requires this to work properly?

Reply:
Maybe this should just be a note about how this function really just performs a particular kind of roll, so its input needs to be structured like fftshift's output?

Reply:
>Isn't it more like the function requires this to work properly?

No, the Nyquist term will always exist in the input. This note is describes where the Nyquist term gets placed in the return value.

Because of the sampling theorem, the Nyquist frequency term in both positive and negative frequencies are identical. So, what changes is whether you include the Nyquist term at the start (treating it as negative), or at the end (treating it as positive). Mathematically, these are equally valid choices.

Reply:
I understand that. But isn't the point that if the term was positive then the operation performed by fftshift wouldn't reorder it properly?

Reply:
>But isn't the point that if the term was positive then the operation performed by fftshift wouldn't reorder it properly?

I'm not quite sure what you mean. The point it that `fftshift` puts the Nyquist term in the left-most position, which corresponds to a negative frequency.

Reply:
Right. Actually being explicit that fftshift just always puts that term at the zero index position would be good. The point I'm trying to get at is the function just always does that, regardless of the actual values. And so for the function fulfill its conceptual purpose, that value should be negative.

Does that make sense?

We could also have a brief VC to discuss this. Sometimes text isn't the greatest way to communicate.

Reply:
>Right. Actually being explicit that fftshift just always puts that term at the zero index position would be good. 

I updated the wording to "fftshift always puts the Nyquist term at the 0-index; equivalent to the negative frequency. This is the same convention used by fftfreq."

>The point I'm trying to get at is the function just always does that, regardless of the actual values. And so for the function fulfill its conceptual purpose, that value should be negative.

I don't understand this point. The FFT gives amplitudes of waves, not their frequencies. The value of the Nyquist term can be anything and does not depend on your convention for the sign of the frequency.

Reply:
I get your point. Maybe we should briefly vc about it? Slack me sometime that's convenient for you?